cmd/otel-allocator/target/discovery.go

```diff
@@ -105,43 +126,110 @@ func (m *Discoverer) ApplyConfig(source allocatorWatcher.EventSource, scrapeConf
 	return m.manager.ApplyConfig(discoveryCfg)
 }
 
-func (m *Discoverer) Watch(fn func(targets map[string]*Item)) error {
-	labelsBuilder := labels.NewBuilder(labels.EmptyLabels())
+func (m *Discoverer) Watch() error {
+	err := m.run(m.manager.SyncCh())
+	if err != nil {
+		m.log.Error(err, "Service Discovery watch event failed")
+		return err
+	}
+	<-m.close
+	m.log.Info("Service Discovery watch event stopped: discovery manager closed")
+	return nil
+}
+
+func (m *Discoverer) UpdateTsets(tsets map[string][]*targetgroup.Group) {
+	m.mtxScrape.Lock()
+	m.targetSets = tsets
+	m.mtxScrape.Unlock()
+}
+
+func (m *Discoverer) reloader() {
+	reloadIntervalDuration := model.Duration(5 * time.Second)
+	ticker := time.NewTicker(time.Duration(reloadIntervalDuration))
+
+	defer ticker.Stop()
+
 	for {
 		select {
 		case <-m.close:
-			m.log.Info("Service Discovery watch event stopped: discovery manager closed")
-			return nil
-		case tsets := <-m.manager.SyncCh():
-			m.ProcessTargets(labelsBuilder, tsets, fn)
+			return
+		case <-ticker.C:
+			select {
+			case <-m.triggerReload:
+				m.Reload()
+			case <-m.close:
+				return
+			}
 		}
 	}
 }
 
-func (m *Discoverer) ProcessTargets(builder *labels.Builder, tsets map[string][]*targetgroup.Group, fn func(targets map[string]*Item)) {
+func (m *Discoverer) Reload() {
+	m.mtxScrape.Lock()
+	var wg sync.WaitGroup
 	targets := map[string]*Item{}
+	timer := prometheus.NewTimer(processTargetsDuration)
+	defer timer.ObserveDuration()
 
-	for jobName, tgs := range tsets {
-		var count float64 = 0
-		for _, tg := range tgs {
-			builder.Reset(labels.EmptyLabels())
-			for ln, lv := range tg.Labels {
+	for jobName, groups := range m.targetSets {
+		wg.Add(1)
+		// Run the sync in parallel as these take a while and at high load can't catch up.
+		go func(jobName string, groups []*targetgroup.Group) {
+			m.processTargetGroups(jobName, groups, targets)
+			wg.Done()
```

Comment:
If we change`processTargetGroups` to return a map we will still need a synchronisation system since it's executed inside a go routine and copy from the returned map to a global run we need to lock it to avoid race conditions.

Maybe we can try to use a sync map, wdyt @swiatekm ?

Reply:
> If we change`processTargetGroups` to return a map we will still need a synchronisation system since it's executed inside a go routine and copy from the returned map to a global run we need to lock it to avoid race conditions.
> 
> Maybe we can try to use a sync map, wdyt @swiatekm ?


Yeah, but we'd lock once per job instead of once per target, and the concurrent calls to `processTargetGroups` could run without any locking. I'm primarily concerned with making the concurrency more visible. `sync.Map` sounds overkill, as we only construct the map concurrently, and afterwards it's not modified other than by serial relabelling.